src/Typedefs/TermWrite.idr

```diff
@@ -10,135 +13,221 @@ import Data.ByteArray
+              {args : Vect n Type} -> HasGenericWriters target args -> ApplyVect constr args -> target
+magicLookup Here (s :: sp) = s
+magicLookup (There e) (s :: sp) = magicLookup e sp
+
+||| Returns the index and the type of a sum from an array of TDefs
+injectionInv : (ts : Vect (2 + k) (TDefR n)) -> Tnary' sp tvars ts Either -> (i : Fin (2 + k) ** Ty' sp tvars (index i ts))
 injectionInv [a,b] (Left x) = (0 ** x)
 injectionInv [a,b] (Right y) = (1 ** y)
 injectionInv (a::b::c::tds) (Left x) = (0 ** x)
 injectionInv (a::b::c::tds) (Right y) =
   let (i' ** y') = injectionInv (b::c::tds) y in (FS i' ** y')
 
-serializeBinary : (t : TDefR n) -> (ts : Vect n (a ** Serialiser a)) ->
-                  Serialiser (Ty (map DPair.fst ts) t)
-serializeBinary T0 ts x impossible
-serializeBinary T1 ts x = empty
-serializeBinary (TSum {k = k} tds) ts x =
-   let (i ** x') = injectionInv tds x in
-   (serializeInt i) ++ (assert_total $ serializeBinary (index i tds) ts x')
-serializeBinary (TProd [a, b]) ts (x, y) =
-  (serializeBinary a ts x) ++ (serializeBinary b ts y)
-serializeBinary (TProd (a::b::c::tds)) ts (x, y) =
-  (serializeBinary a ts x) ++ (serializeBinary (TProd (b::c::tds))) ts y
-serializeBinary (TMu tds) ts (Inn x) =
-  assert_total $  serializeBinary (args tds) ((Mu (map DPair.fst ts) (args tds) ** serializeBinary (TMu tds) ts)::ts) x
-serializeBinary (TVar FZ) (t::ts) x = snd t x
-serializeBinary {n = S (S n')} (TVar (FS i)) (t::ts) x =
-  serializeBinary {n = (S n')} (TVar i) ts x
-serializeBinary (RRef FZ) (t::ts) x = snd t x
-serializeBinary {n = S (S n')} (RRef (FS i)) (t::ts) x =
-  serializeBinary {n = (S n')} (RRef i) ts x
-serializeBinary (TApp (TName n d) xs) ts x =
-  assert_total $ serializeBinary (ap d xs) ts (convertTy {n=n} {v=map DPair.fst ts} {def=d} {xs=xs} x)
-
-serializeBinaryClosed : (t : TDefR 0) -> Serialiser (Ty [] t)
-serializeBinaryClosed t = serializeBinary t []
-
-
-HasJSONWriters : Vect n Type -> Type
-HasJSONWriters = HasGenericWriters JSON
+getVariable : {ts : Vect (S n) Type} -> (i : Fin (S n)) -> (ws : HasGenericWriters target ts) -> index i ts -> target
+getVariable FZ (f :: z) x {ts = (y :: xs)} = f x
+getVariable (FS FZ) (_ :: f :: ws) x = f x
+getVariable (FS (FS y)) (w :: ws) x = getVariable (FS y) ws x
+
+||| Given a vector of TDef n and n types to fill its holes, return the vector of types correpondng to those tdefs
+FromTDefToTy : {n : Nat} -> SpecialList l -> Vect n Type -> Vect k (TDefR n) -> Vect k Type
+FromTDefToTy sp types vs = map (Ty' sp types) vs
+
+||| Magically generate the writers for a vector of TDefs in argument position
+|||
+||| For some reason this couldn't be placed in a where-clause inside `serialise`. The lookup for
+||| instances of `TDefSeraliser` would get confused and say that there aren't any suitable instances
+||| for `serialiser`. That's why we simply pass it as a partially applied function
+magicWriters : {target : Type} -> {sp : SpecialList l} ->
```

Comment:
What's so magical about it? Looks like a `map` in disguise :)

Reply:
renamed to `makeWriters`